File: frontend/src/types/analysis.ts
export interface PerceptualHash {
  hash_algorithm: string;
  hash_value: string;
}

export interface EvidenceResultData {
  analyzer_name: string;
  analyzer_version: string;
  status: string;
  confidence: number | null;
  findings: Record<string, unknown>[] | null;
  limitations: string[] | null;
  duration_ms: number | null;
  created_at: string;
}

export interface AnalysisSummary {
  analysis_id: string;
  status: string;
  classification: string | null;
  overall_confidence: number | null;
  original_sha256: string;
  file_size: number;
  mime_type: string;
  detected_format: string;
  width: number | null;
  height: number | null;
  created_at: string;
  completed_at: string | null;
}

export interface AnalysisDetail extends AnalysisSummary {
  bit_depth: number | null;
  color_mode: string | null;
  color_profile: string | null;
  compression_info: Record<string, unknown> | null;
  analysis_version: string;
  software_version: string;
  model_versions: Record<string, string> | null;
  error_message: string | null;
  evidence_results: EvidenceResultData[];
}

export interface ImageInfo {
  sha256: string;
  file_size: number;
  mime_type: string;
  format: string;
  width: number | null;
  height: number | null;
  bit_depth: number | null;
  color_mode: string | null;
  color_profile: string | null;
  compression_info: Record<string, unknown> | null;
  perceptual_hashes: PerceptualHash[];
}

export interface C2PAInfo {
  present: boolean;
  valid: boolean | null;
  trusted: boolean | null;
  signer: string | null;
  claim_generator: string | null;
  actions: Record<string, unknown>[];
  ingredients: Record<string, unknown>[];
  validation_errors: string[];
  trust_signals: string[];
}

export interface WatermarkResult {
  provider: string;
  watermark_type: string;
  detected: boolean;
  confidence: number;
  detector_version: string;
  limitations: string[];
  evidence: Record<string, unknown>;
}

export interface ProvenanceSection {
  c2pa: C2PAInfo;
  metadata: Record<string, unknown>;
  watermarks: WatermarkResult[];
}

export interface ForensicsSection {
  compression: Record<string, unknown>;
  ela: Record<string, unknown>;
  noise: Record<string, unknown>;
  resampling: Record<string, unknown>;
  copy_move: Record<string, unknown>;
  statistics: Record<string, unknown>;
}

export interface MLSection {
  ai_generated: number | null;
  ai_edited: number | null;
  conventional_edit: number | null;
  provider_prediction: string | null;
  provider_confidence: number | null;
  model_name: string | null;
  model_version: string | null;
  status: string;
}

export interface EvidenceItem {
  source: string;
  category: string;
  description: string;
  confidence: number | null;
  supports: string | null;
}

export interface ForensicReport {
  analysis_id: string;
  classification: string | null;
  overall_confidence: number | null;
  image_info: ImageInfo;
  provenance: ProvenanceSection;
  forensics: ForensicsSection;
  ml: MLSection;
  regions: Record<string, unknown>[];
  evidence_chain: EvidenceItem[];
  limitations: string[];
}

export interface HealthResponse {
  status: string;
  version: string;
  analyzers: Record<string, string>;
}

export type ClassificationType =
  | "verified_ai_provenance"
  | "likely_ai_generated"
  | "likely_ai_edited"
  | "conventionally_edited"
  | "likely_original"
  | "mixed_provenance"
  | "inconclusive";

export const CLASSIFICATION_LABELS: Record<ClassificationType, string> = {
  verified_ai_provenance: "Verified AI Provenance",
  likely_ai_generated: "Likely AI-Generated",
  likely_ai_edited: "Likely AI-Edited",
  conventionally_edited: "Conventionally Edited",
  likely_original: "Likely Original",
  mixed_provenance: "Mixed Provenance",
  inconclusive: "Inconclusive",
};

File: backend/app/schemas/analysis.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class EvidenceResultOut(BaseModel):
    analyzer_name: str
    analyzer_version: str
    status: str
    confidence: float | None = None
    findings: list[dict[str, Any]] | None = None
    limitations: list[str] | None = None
    duration_ms: int | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class PerceptualHashOut(BaseModel):
    hash_algorithm: str
    hash_value: str

    model_config = {"from_attributes": True}


class AnalysisSummary(BaseModel):
    analysis_id: UUID
    status: str
    classification: str | None = None
    overall_confidence: float | None = None
    original_sha256: str
    file_size: int
    mime_type: str
    detected_format: str
    width: int | None = None
    height: int | None = None
    created_at: datetime
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}


class AnalysisDetail(AnalysisSummary):
    bit_depth: int | None = None
    color_mode: str | None = None
    color_profile: str | None = None
    compression_info: dict[str, Any] | None = None
    analysis_version: str
    software_version: str
    model_versions: dict[str, str] | None = None
    error_message: str | None = None
    evidence_results: list[EvidenceResultOut] = Field(default_factory=list)


class ForensicReport(BaseModel):
    analysis_id: UUID
    classification: str | None
    overall_confidence: float | None

    image_info: ImageInfo
    provenance: ProvenanceSection
    forensics: ForensicsSection
    ml: MLSection
    regions: list[dict[str, Any]] = Field(default_factory=list)
    evidence_chain: list[EvidenceItem] = Field(default_factory=list)
    limitations: list[str] = Field(default_factory=list)


class ImageInfo(BaseModel):
    sha256: str
    file_size: int
    mime_type: str
    format: str
    width: int | None = None
    height: int | None = None
    bit_depth: int | None = None
    color_mode: str | None = None
    color_profile: str | None = None
    compression_info: dict[str, Any] | None = None
    perceptual_hashes: list[PerceptualHashOut] = Field(default_factory=list)


class C2PAInfo(BaseModel):
    present: bool = False
    valid: bool | None = None
    trusted: bool | None = None
    signer: str | None = None
    claim_generator: str | None = None
    actions: list[dict[str, Any]] = Field(default_factory=list)
    ingredients: list[dict[str, Any]] = Field(default_factory=list)
    validation_errors: list[str] = Field(default_factory=list)
    trust_signals: list[str] = Field(default_factory=list)


class WatermarkResult(BaseModel):
    provider: str
    watermark_type: str
    detected: bool
    confidence: float
    detector_version: str
    limitations: list[str] = Field(default_factory=list)
    evidence: dict[str, Any] = Field(default_factory=dict)


class ProvenanceSection(BaseModel):
    c2pa: C2PAInfo = Field(default_factory=C2PAInfo)
    metadata: dict[str, Any] = Field(default_factory=dict)
    watermarks: list[WatermarkResult] = Field(default_factory=list)


class ForensicsSection(BaseModel):
    compression: dict[str, Any] = Field(default_factory=dict)
    ela: dict[str, Any] = Field(default_factory=dict)
    noise: dict[str, Any] = Field(default_factory=dict)
    resampling: dict[str, Any] = Field(default_factory=dict)
    copy_move: dict[str, Any] = Field(default_factory=dict)
    statistics: dict[str, Any] = Field(default_factory=dict)


class MLSection(BaseModel):
    ai_generated: float | None = None
    ai_edited: float | None = None
    conventional_edit: float | None = None
    provider_prediction: str | None = None
    provider_confidence: float | None = None
    model_name: str | None = None
    model_version: str | None = None
    status: str = "not_implemented"


class EvidenceItem(BaseModel):
    source: str
    category: str
    description: str
    confidence: float | None = None
    supports: str | None = None


class HealthResponse(BaseModel):
    status: str = "ok"
    version: str
    analyzers: dict[str, str] = Field(default_factory=dict)


class VersionResponse(BaseModel):
    version: str
    api_version: str
    analyzers: dict[str, str] = Field(default_factory=dict)
    ml_models: dict[str, str] = Field(default_factory=dict)


ForensicReport.model_rebuild()
